lib/Scheduling/TestPasses.cpp

```diff
@@ -361,6 +367,49 @@ void TestSimplexSchedulerPass::runOnFunction() {
   llvm_unreachable("Unsupported scheduling problem");
 }
 
+//===----------------------------------------------------------------------===//
+// ComponentLibrary
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct TestComponentLibraryPass
+    : public PassWrapper<TestComponentLibraryPass, FunctionPass> {
+  TestComponentLibraryPass() = default;
+  TestComponentLibraryPass(const TestComponentLibraryPass &) {}
+  void runOnFunction() override;
+  StringRef getArgument() const override { return "test-component-library"; }
+  StringRef getDescription() const override {
+    return "Test the scheduling and lowering of ComponentLibrary";
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<comb::CombDialect>();
+  }
+};
+
+// Example component library that supports combinational adders and 3-cycle
+// multipliers. Dialects that want to support scheduling their primitives will
+// define something similar.
+struct TestComponentLibrary : ComponentLibrary {
```

Comment:
Cool! How do you foresee support for libraries that support multiple variations of a given Operation, e.g. a 3-cycle multiplier and a 0-cycle multiplier?

Reply:
I'd imagine there could be another parameter to `addComponent`, like "benefit" or something like that. We'd have to update the datastructure to store a list of ComponentData structs for each operation, and select the one with the highest benefit. Something along those lines.